File: backend/app/schemas/backup.py
"""Backup request and response schemas."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class BackupBase(BaseModel):
    """Base backup schema."""
    backup_type: str = Field(default="full", pattern="^(full|incremental|database_only|files_only)$")
    includes_database: bool = True
    includes_files: bool = True
    includes_config: bool = False


class BackupCreate(BackupBase):
    """Schema for creating a new backup."""
    backup_path: Optional[str] = None


class BackupInDB(BackupBase):
    """Schema for backup stored in database."""
    id: int
    filename: str
    filepath: str
    size_bytes: int
    status: str
    created_at: datetime
    completed_at: Optional[datetime] = None
    creator_id: int
    error_message: Optional[str] = None

    model_config = {"from_attributes": True}


class BackupResponse(BackupInDB):
    """Schema for backup response to client."""
    size_mb: float = Field(description="Size in megabytes for display")

    @classmethod
    def from_db(cls, backup: BackupInDB) -> "BackupResponse":
        """Convert database model to response schema."""
        return cls(
            **backup.model_dump(),
            size_mb=round(backup.size_bytes / (1024 * 1024), 2)
        )


class BackupListResponse(BaseModel):
    """Schema for listing backups."""
    backups: list[BackupResponse]
    total_size_bytes: int
    total_size_mb: float


class BackupRestoreRequest(BaseModel):
    """Schema for restore request."""
    backup_id: int
    restore_database: bool = True
    restore_files: bool = True
    restore_config: bool = False
    confirm: bool = Field(description="User must confirm restore operation")


class BackupRestoreResponse(BaseModel):
    """Schema for restore response."""
    success: bool
    message: str
    backup_id: int
    restored_at: datetime

File: client/src/api/backup.ts
/**
 * API client for backup functionality
 */

import { apiClient, memoizedApiRequest } from '../lib/api';

export interface Backup {
  id: number;
  filename: string;
  filepath: string;
  size_bytes: number;
  size_mb: number;
  backup_type: 'full' | 'incremental' | 'database_only' | 'files_only';
  status: 'in_progress' | 'completed' | 'failed';
  created_at: string;
  completed_at: string | null;
  creator_id: number;
  error_message: string | null;
  includes_database: boolean;
  includes_files: boolean;
  includes_config: boolean;
}

export interface CreateBackupRequest {
  backup_type?: 'full' | 'incremental' | 'database_only' | 'files_only';
  includes_database?: boolean;
  includes_files?: boolean;
  includes_config?: boolean;
  backup_path?: string;
}

export interface BackupListResponse {
  backups: Backup[];
  total_size_bytes: number;
  total_size_mb: number;
}

export interface RestoreBackupRequest {
  backup_id: number;
  restore_database?: boolean;
  restore_files?: boolean;
  restore_config?: boolean;
  confirm: boolean;
}

export interface RestoreBackupResponse {
  success: boolean;
  message: string;
  backup_id: number;
  restored_at: string;
}

/**
 * Create a new system backup
 */
export async function createBackup(request: CreateBackupRequest = {}): Promise<Backup> {
  const response = await apiClient.post<Backup>('/api/backups/', request);
  return response.data;
}

/**
 * List all backups
 */
export async function listBackups(): Promise<BackupListResponse> {
  // Memoized: Backups werden für 60s gecached
  return memoizedApiRequest<BackupListResponse>('/api/backups/');
}

/**
 * Get backup details by ID
 */
export async function getBackup(backupId: number): Promise<Backup> {
  // Memoized: Backup-Details werden für 60s gecached
  return memoizedApiRequest<Backup>(`/api/backups/${backupId}/`);
}

/**
 * Delete a backup
 */
export async function deleteBackup(backupId: number): Promise<void> {
  await apiClient.delete(`/api/backups/${backupId}`);
}

/**
 * Restore system from backup
 */
export async function restoreBackup(request: RestoreBackupRequest): Promise<RestoreBackupResponse> {
  const token = localStorage.getItem('token');
  const response = await apiClient.post<RestoreBackupResponse>(
    `/api/backups/${request.backup_id}/restore`,
    request,
    {
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data;
}

/**
 * Download backup file
 */
export function getBackupDownloadUrl(backupId: number): string {
  const baseUrl = import.meta.env.VITE_API_URL || 'http://localhost:8000';
  const token = localStorage.getItem('token');
  return `${baseUrl}/api/backups/${backupId}/download?token=${token}`;
}

/**
 * Download backup file (triggers browser download)
 */
export async function downloadBackup(backupId: number, filename: string): Promise<void> {
  const url = getBackupDownloadUrl(backupId);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
}
